TrackInfo/MediaSamplesBlock do not allow for protected content.

jyavenard opened the pull request:
#### 795999704c17ab8b2e2d3d61465c52255b216c29
<pre>
TrackInfo/MediaSamplesBlock do not allow for protected content.
<a href="https://bugs.webkit.org/show_bug.cgi?id=300875">https://bugs.webkit.org/show_bug.cgi?id=300875</a>
<a href="https://rdar.apple.com/162759126">rdar://162759126</a>

Reviewed by Youenn Fablet.

We add preliminary support for FairPlay encrypted content and ability to
convert an encrypted CMSampleBuffer to/from a MediaSamplesBlock.

Manually tested with MediaSourceUseRemoteAudioVideoRenderer set to true
and playing Netflix and Disney videos.
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.h:
* Source/WebCore/platform/MediaSamplesBlock.h:
* Source/WebCore/platform/TrackInfo.h:
(WebCore::TrackInfo::operator== const):
(WebCore::TrackInfo::TrackInfo):
(WebCore::VideoInfo::create):
(WebCore::VideoInfo::VideoInfo):
(WebCore::AudioInfo::create):
(WebCore::AudioInfo::AudioInfo):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSampleAVFObjC.mm:
(WebCore::MediaSampleAVFObjC::commonInit): Move getKeyIDs into its own method in CMUtilities.
* Source/WebCore/platform/graphics/cocoa/CMUtilities.h:
* Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
(WebCore::cfStringToFourCC):
(WebCore::cfStringFromFourCC):
(WebCore::createExtensionAtomsDictionary):
(WebCore::setEncryptionInfo):
(WebCore::createAudioFormatDescription):
(WebCore::createFormatDescriptionFromTrackInfo):
(WebCore::createAudioInfoFromFormatDescription):
(WebCore::createVideoInfoFromFormatDescription):
(WebCore::toCMSampleBuffer):
(WebCore::samplesBlockFromCMSampleBuffer):
(WebCore::getKeyIDs):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/302086@main">https://commits.webkit.org/302086@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/42f1151bef062be7356ffd02f6e750e0ac561ed1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127996 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/270 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38822 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135367 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/79511 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/0fdfa25c-674b-4ee7-a693-13d2270b6a4b) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129868 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/165 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/147 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/97437 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/79511 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b15193a4-4699-4a97-b10c-79d9fc6b3d2d) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130944 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/93 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114657 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78005 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/85a488e4-cd5e-406b-bf38-41347fe8b2fa) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/127347 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/99 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32761 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78676 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/108458 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33240 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137850 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/135 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/125 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/105965 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/160 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111007 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105702 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26939 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/102 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/29562 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52292 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/181 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/61644 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/104 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/180 "Built successfully") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/145 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/TrackInfo.h:
```diff
@@ -66,13 +71,25 @@ struct TrackInfo : public ThreadSafeRefCounted<TrackInfo> {
+#if ENABLE(ENCRYPTED_MEDIA)
+    using EncryptionData = std::pair<EncryptionBoxType, Ref<SharedBuffer>>;
+    std::optional<EncryptionData> encryptionData { };
+    std::optional<FourCC> encryptionOriginalFormat { };
```
Do we need `{ }`?

jyavenard replied:
Yes. Otherwise when initialising the struct, we must do every member.

youennf commented on Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
```diff
@@ -636,6 +817,41 @@ void attachColorSpaceToPixelBuffer(const PlatformVideoColorSpace& colorSpace, CV
+
+        ISOTrackEncryptionBox trackEncryptionBox;
+        auto trackEncryptionView = JSC::DataView::create(WTFMove(trackEncryptionDataBuffer), 0, length);
+        if (!trackEncryptionBox.parseWithoutTypeAndSize(trackEncryptionView))
```
It seems we should have ISOTrackEncryptionBox take a std::span instead of a DataView.

jyavenard replied:
I just.moved that code

youennf commented on Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
```diff
@@ -636,6 +817,41 @@ void attachColorSpaceToPixelBuffer(const PlatformVideoColorSpace& colorSpace, CV
+    if (static_cast<CFDataRef>(PAL::CMFormatDescriptionGetExtension(description, CFSTR("TransportStreamEncryptionInitData")))) {
+        // AVStreamDataParser will attach a JSON transport stream encryption
+        // description object to each sample. Use a static keyID in this case
+        // as MPEG2-TS encryption dose not specify a particular keyID in the
```
```suggestion
        // as MPEG2-TS encryption does not specify a particular keyID in the
```

youennf commented on Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
```diff
@@ -636,6 +817,41 @@ void attachColorSpaceToPixelBuffer(const PlatformVideoColorSpace& colorSpace, CV
 }
 
+#if ENABLE(ENCRYPTED_MEDIA) && HAVE(AVCONTENTKEYSESSION)
+Vector<Ref<SharedBuffer>> getKeyIDs(CMFormatDescriptionRef description)
```
It seems we only return one KeyID, will that change in the future?

jyavenard replied:
The code it calls can return more than one in its structure. But as it is, I don't know

youennf commented on Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
```diff
@@ -424,18 +583,40 @@ static CFStringRef convertToCMYCbCRMatrix(PlatformVideoMatrixCoefficients coeffi
+        RetainPtr attachmentsArray = PAL::CMSampleBufferGetSampleAttachmentsArray(sample->sampleBuffer(), false);
+        if (attachmentsArray && CFArrayGetCount(attachmentsArray.get()) > 0) {
+            if (RetainPtr attachmentsDictionary = dynamic_cf_cast<CFMutableDictionaryRef>(CFArrayGetValueAtIndex(attachmentsArray.get(), 0))) {
+                if (RetainPtr number = dynamic_cf_cast<CFNumberRef>(CFDictionaryGetValue(attachmentsDictionary.get(), CFSTR("BytesOfClearDataCount") /* PAL::kCMSampleAttachmentKey_BytesOfClearDataCount */)))
```
Why not using kCMSampleAttachmentKey_BytesOfClearDataCount, ditto below?

jyavenard replied:
SPI

youennf commented on Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
```diff
@@ -401,6 +537,29 @@ static CFStringRef convertToCMYCbCRMatrix(PlatformVideoMatrixCoefficients coeffi
+
+    RetainPtr attachmentsArray = PAL::CMSampleBufferGetSampleAttachmentsArray(rawSampleBuffer, false);
+    if (!attachmentsArray || static_cast<size_t>(CFArrayGetCount(attachmentsArray.get())) < samples.size())
+        return adoptCF(rawSampleBuffer);
```
Should we log this error case?

youennf commented on Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
```diff
@@ -192,37 +286,50 @@ static CFStringRef convertToCMYCbCRMatrix(PlatformVideoMatrixCoefficients coeffi
-        RetainPtr<CFStringRef> keyName;
-        if (!videoInfo.boxType.isEmpty())
-            keyName = videoInfo.boxType.createCFString();
+        if (videoInfo.boxType.value)
```
I would be tempted to add a VideoInfo method or function to compute boxType based on boxType if available and codecName otherwise.
Then we could write it as:
`configurations.append({ videoInfo.computeBoxType(), atomData.releaseNonNull() });`

Maybe VideoInfo::boxType should be made private, as a way to progressively move VideoInfo to a class and not a struct.

jyavenard replied:
idea is good, but it's not a very portable approach unfortunately. It's AVFoundation/CoreMedia specific.

jyavenard committed:
TrackInfo/MediaSamplesBlock do not allow for protected content.
https://bugs.webkit.org/show_bug.cgi?id=300875
rdar://162759126

Reviewed by Youenn Fablet.

We add preliminary support for FairPlay encrypted content and ability to
convert an encrypted CMSampleBuffer to/from a MediaSamplesBlock.

Manually tested with MediaSourceUseRemoteAudioVideoRenderer set to true
and playing Netflix and Disney videos.
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.h:
* Source/WebCore/platform/MediaSamplesBlock.h:
* Source/WebCore/platform/TrackInfo.h:
(WebCore::TrackInfo::operator== const):
(WebCore::TrackInfo::TrackInfo):
(WebCore::VideoInfo::create):
(WebCore::VideoInfo::VideoInfo):
(WebCore::AudioInfo::create):
(WebCore::AudioInfo::AudioInfo):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSampleAVFObjC.mm:
(WebCore::MediaSampleAVFObjC::commonInit): Move getKeyIDs into its own method in CMUtilities.
* Source/WebCore/platform/graphics/cocoa/CMUtilities.h:
* Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
(WebCore::cfStringToFourCC):
(WebCore::cfStringFromFourCC):
(WebCore::createExtensionAtomsDictionary):
(WebCore::setEncryptionInfo):
(WebCore::createAudioFormatDescription):
(WebCore::createFormatDescriptionFromTrackInfo):
(WebCore::createAudioInfoFromFormatDescription):
(WebCore::createVideoInfoFromFormatDescription):
(WebCore::toCMSampleBuffer):
(WebCore::samplesBlockFromCMSampleBuffer):
(WebCore::getKeyIDs):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/302086@main
